### PR TITLE
Add support for bulk memory operations and reference types

### PR DESCRIPTION
**Pull request by EricSL**
This is tested for disassembling the entire file in one chunk.
More work is needed to support streaming disassembly or emitting
Wasm. Test files generated with wabt/wat2wasm built at head based
on round-trip tests.

### CODE REVIEW

**Comment by EricSL** [comment]
Yury,

Is it important to support these operations in WasmEmitter as well? Also, my changes would I assume be broken in partial disassembly, but I wasn't able to figure out how to test that.

**Comment by yurydelendik** [comment]
> Is it important to support these operations in WasmEmitter as well?

No, I'm not aware of any WasmEmitter user atm.

**Comment by EricSL** [comment]
> Overall looks good. Though it will be nice to add different state for `PASSIVE_ELEMENT_SECTION_ENTRY` and remove _segmentFlags as state. (I previously trapped on this with wasmparser.rs)

This could work for data sections but not for element sections. Note that there are multiple subsequent if conditions based on the flags:

line 1122: if (this._segmentFlags & (SegmentFlags.IsPassive | SegmentFlags.HasTableIndex)) {
line 1136: if (this._segmentFlags & SegmentFlags.FunctionsAsElements) {
line 2253: if (this._segmentFlags & SegmentFlags.IsPassive) {

Also note that there may be an initializer expression, which involves transitioning through a few states, and after that completes we have to condition on the segment flags.

**Comment by yurydelendik** [comment]
> This could work for data sections but not for element sections. Note that there are multiple subsequent if conditions based on the flags:
> 
> line 1122: if (this._segmentFlags & (SegmentFlags.IsPassive | SegmentFlags.HasTableIndex)) {
> line 1136: if (this._segmentFlags & SegmentFlags.FunctionsAsElements) {
> line 2253: if (this._segmentFlags & SegmentFlags.IsPassive) {
> 
 
Right, active/passive states present for element and data. In all of the cases `_segmentFlags` selects a state, which defines condition for the if's you pointed above. It might worth to duplicate functions code instead of adding `if` checks.

**Comment by EricSL** [comment]
I don't understand what you are asking me to do here. These are different conditions, so it's not a matter of just one more state. And in the non-passive case, there is an initializer expression in the body. Currently when completing an initializer expression the subsequent state is determined by what kind of section we are in, but with this change after completing the initializer expression we need to remember whether the FunctionsAsElements flag and the HasTableIndex flag were set.

**Comment by yurydelendik** [comment]
> the subsequent state is determined by what kind of section we are in, but with this change after completing the initializer expression we need to remember whether the FunctionsAsElements flag and the HasTableIndex flag were set.

I'm looking at https://github.com/WebAssembly/bulk-memory-operations/blob/master/proposals/bulk-memory-operations/Overview.md#design at the "Another way of looking at it"-tables. There are two distinct states with different read elements logic: Passive and Active.

Also, I would like to keep wasmparser.js and wasmparser.rs projects "in-sync", so further contributions be less painful.

**Comment by EricSL** [comment]
Yes, that document confused me during implementation, as it is out of date. My code is consistent with what v8 and also wabt (at head) implement. Also, the detailed documentation has been updated, just not the overview:

https://github.com/WebAssembly/bulk-memory-operations/blob/master/document/core/binary/modules.rst#element-section

> The initial byte can be interpreted as a bitfield. Bit 0 indicates a passive segment, bit 1 indicates the presence of an explicit table index for an active segment, bit 2 indicates the use of element type and element expressions instead of element kind and element indices.

**Comment by EricSL** [comment]
The overview document has been updated to match.

File: src/WasmDis.ts [hunk 1]
```diff
@@ -18,8 +18,9 @@ import {
   IImportEntry, IOperatorInformation, Type, OperatorCode, OperatorCodeNames, Int64,
   ITableType, IMemoryType, IGlobalType, IResizableLimits, IDataSegmentBody,
   IGlobalVariable, IElementSegment, IElementSegmentBody, ISectionInformation,
-  IStartEntry, bytesToString, INameEntry, NameType, IFunctionNameEntry, INaming
-} from './WasmParser';
+  IStartEntry, bytesToString, INameEntry, NameType, IFunctionNameEntry, INaming,
+  NULL_FUNCTION_INDEX
+} from './WasmParser.js';
```
Comment (yurydelendik) [inline src/WasmDis.ts]: `from './WasmParser.js'` -- can we still use from './WasmParser'?

**Comment by EricSL** [comment]
It was compiling to code which imported without an extension, which our
packaging system didn't like. Is there a way to get Typescript to add this
extension?

**Comment by yurydelendik** [comment]
The wasmparser is published in npm registry -- we translate TS to ES Modules via `npm run build`. It is used as JS library in our end products.

>  Is there a way to get Typescript to add this extension?

Not sure

**Comment by yurydelendik** [comment]
Thank you for the patch!

**Pull request merged**